Distributions.ecl

```diff
@@ -0,0 +1,407 @@
+//Normal, Student and Chi Squared distributions for tests
+IMPORT ML_Core.Constants AS Core_Const;
+IMPORT ML_Core.Math AS Math;
+
+EXPORT Distributions := MODULE
+  /**
+   * Cumulative Distribution of the standard normal distribution, the
+   *probability that a normal random variable will be smaller than x
+   * standard deviations above or below the mean.
+   * Taken from C/C++ Mathematical Algorithms for Scientists and
+   *Engineers, n. Shammas, McGraw-Hill, 1995
+   * @param x the number of standard deviations
+   * @returns probability of exceeding x.
+   */
+  EXPORT REAL8 Normal_CDF(REAL8 x) := FUNCTION
+    b1 := 0.319381530;
+    b2 :=-0.356563782;
+    b3 := 1.781477937;
+    b4 :=-1.821255978;
+    b5 := 1.330274429;
+    xp1 := 1.0 / (1.0 + 0.2316419*ABS(x));
+    xp2 := xp1*xp1;
+    xp3 := xp2*xp1;
+    xp4 := xp3*xp1;
+    xp5 := xp4*xp1;
+    term_sum := xp1*b1 + xp2*b2 + xp3*b3 + xp4*b4 + xp5*b5;
+    recip_sqrt_2_pi := 0.39894228040143;
+    rslt0 := exp(-x*x/2) * recip_sqrt_2_pi * term_sum;
+    rslt := IF(x>=0.0, 1-rslt0, rslt0);
+    RETURN rslt;
+  END;
+
+  /**
+   * Normal Distribution Percentage Point Function.
+   * Translated from C/C++ Mathematical Algorithms for Scientists and
+   * Engineers, N. Shammas, McGraw-Hill, 1995
+   * @param x probability
+   * @returns number of standard deviations from the mean
+   */
+  EXPORT REAL8 Normal_PPF(REAL8 x) := FUNCTION
+    p0 := -0.322232431088;
+    p1 := -1.0;
+    p2 := -0.342242088547;
+    p3 := -0.204231210245e-1;
+    p4 := -0.453642210148e-4;
+    q0 :=  0.993484626060e-1;
+    q1 :=  0.588581570495;
+    q2 :=  0.531103462366;
+    q3 :=  0.103537752850;
+    q4 :=  0.38560700634e-2;
+    x0 := MAP(x<=0.0 => 0.0001, x>=1.0 => 0.9999, x);
+    xp := SQRT(-2.0 * LN(IF(x0<=0.5, x0, 1-x0)));
+    numerator := ((((p4*xp+p3)*xp+p2)*xp+p1)*xp+p0);
+    denominator := ((((q4*xp+q3)*xp+q2)*xp+q1)*xp+q0);
+    rslt0 := xp + numerator/denominator;
+    RETURN MAP(x0=0.5  => 0.0,
+               x0<0.5  => -rslt0,
+               rslt0);
+  END;
+
+  // Helper for Students distribution
+  REAL8 Sum_Terms(INTEGER4 df, REAL8 c, REAL8 csq) := BEGINC++
+    #option pure;
+    double work_sum = (df%2==0) ? 1.0  : c;
+    double work_term = (df%2==0) ? 1.0  : c;
+    double start = (df%2==0) ? 2.0  : 3.0;
+    double stop = (double) (df - 2);
+    for (double ai=start; ai<=stop; ai+=2.0) {
+      work_term *= csq * ((ai-1.0)/ai);
+      work_sum  += work_term;
+    }
+    return work_sum;
+  ENDC++;
+  /**
+   * Students t distribution integral evaluated between negative
+   * infinity and x.
+   * Translated from NIST SEL DATAPAC Fortran TCDF.f source
+   * @param x value of the evaluation
+   * @param df degrees of freedom
+   * @returns the probability that a value will be less than the
+   * specified value
+   */
+  EXPORT REAL8 T_CDF(REAL8 x, REAL8 df) := FUNCTION
```

Comment:
Yes.  The underlying distributions support fractional dimensions and df is really the dimension of the space; plus there are applications where the number of observations should be fractional because of near duplicates.